File: app/widgets/library/Library.java
package library;

import java.io.File;
import java.util.Collection;
import java.util.List;
import javafx.concurrent.Task;
import javafx.css.PseudoClass;
import javafx.event.Event;
import javafx.geometry.NodeOrientation;
import javafx.scene.control.Label;
import javafx.scene.control.TableColumn;
import javafx.scene.input.Dragboard;
import javafx.scene.layout.Pane;
import javafx.stage.FileChooser;
import javafx.stage.Window;
import javafx.util.Callback;
import sp.it.pl.audio.Item;
import sp.it.pl.audio.Player;
import sp.it.pl.audio.playlist.PlaylistManager;
import sp.it.pl.audio.tagging.Metadata;
import sp.it.pl.audio.tagging.MetadataGroup;
import sp.it.pl.audio.tagging.MetadataReader;
import sp.it.pl.gui.nodeinfo.TaskInfo;
import sp.it.pl.gui.objects.contextmenu.TableContextMenuR;
import sp.it.pl.gui.objects.table.FilteredTable;
import sp.it.pl.gui.objects.table.ImprovedTable.PojoV;
import sp.it.pl.gui.objects.table.TableColumnInfo;
import sp.it.pl.gui.objects.tablerow.ImprovedTableRow;
import sp.it.pl.layout.widget.Widget;
import sp.it.pl.layout.widget.Widget.Info;
import sp.it.pl.layout.widget.controller.LegacyController;
import sp.it.pl.layout.widget.controller.SimpleController;
import sp.it.pl.layout.widget.controller.io.IsInput;
import sp.it.pl.layout.widget.controller.io.Output;
import sp.it.pl.layout.widget.feature.SongReader;
import sp.it.pl.main.Widgets;
import sp.it.pl.util.access.Vo;
import sp.it.pl.util.access.fieldvalue.ColumnField;
import sp.it.pl.util.animation.Anim;
import sp.it.pl.util.animation.interpolator.ElasticInterpolator;
import sp.it.pl.util.conf.Config;
import sp.it.pl.util.conf.EditMode;
import sp.it.pl.util.conf.IsConfig;
import sp.it.pl.util.file.AudioFileFormat;
import sp.it.pl.util.file.AudioFileFormat.Use;
import sp.it.pl.util.file.FileType;
import sp.it.pl.util.graphics.drag.DragUtil;
import sp.it.pl.util.units.Dur;
import sp.it.pl.util.validation.Constraint;
import sp.it.pl.util.validation.Constraint.FileActor;
import static javafx.scene.control.SelectionMode.MULTIPLE;
import static javafx.scene.control.TableView.UNCONSTRAINED_RESIZE_POLICY;
import static javafx.scene.input.KeyCode.DELETE;
import static javafx.scene.input.KeyCode.ENTER;
import static javafx.scene.input.MouseButton.PRIMARY;
import static javafx.scene.input.TransferMode.COPY;
import static javafx.util.Duration.millis;
import static javafx.util.Duration.seconds;
import static sp.it.pl.audio.tagging.Metadata.Field.RATING;
import static sp.it.pl.audio.tagging.Metadata.Field.TITLE;
import static sp.it.pl.gui.nodeinfo.TableInfo.DEFAULT_TEXT_FACTORY;
import static sp.it.pl.layout.widget.Widget.Group.LIBRARY;
import static sp.it.pl.main.AppBuildersKt.appProgressIndicator;
import static sp.it.pl.main.AppKt.APP;
import static sp.it.pl.util.animation.Anim.Interpolators.reverse;
import static sp.it.pl.util.async.AsyncKt.FX;
import static sp.it.pl.util.async.future.Fut.fut;
import static sp.it.pl.util.file.Util.getCommonRoot;
import static sp.it.pl.util.functional.Util.map;
import static sp.it.pl.util.functional.UtilKt.runnable;
import static sp.it.pl.util.graphics.Util.menuItem;
import static sp.it.pl.util.graphics.UtilKt.setScaleXY;
import static sp.it.pl.util.reactive.UtilKt.maintain;
import static sp.it.pl.util.system.EnvironmentKt.chooseFile;
import static sp.it.pl.util.system.EnvironmentKt.chooseFiles;

@Info(
    author = "Martin Polakovic",
    name = Widgets.SONG_TABLE,
    description = "Provides access to database.",
    howto = "Available actions:\n" +
            "    Item left click : Selects item\n" +
            "    Item right click : Opens context menu\n" +
            "    Item double click : Plays item\n" +
            "    Type : search & filter\n" +
            "    Press ENTER : Plays item\n" +
            "    Press ESC : Clear selection & filter\n" +
            "    Scroll : Scroll table vertically\n" +
            "    Scroll + SHIFT : Scroll table horizontally\n" +
            "    Column drag : swap columns\n" +
            "    Column right click: show column menu\n" +
            "    Click column : Sort - ascending | descending | none\n" +
            "    Click column + SHIFT : Sorts by multiple columns\n" +
            "    Menu bar : Opens additional actions\n",
//    notes = "",
    version = "1",
    year = "2015",
    group = LIBRARY
)
@LegacyController
public class Library extends SimpleController implements SongReader {

    private static final PseudoClass PC_PLAYING = PseudoClass.getPseudoClass("played");
    private static final TableContextMenuR<MetadataGroup> contextMenu = new TableContextMenuR<>();

    private final FilteredTable<Metadata> table = new FilteredTable<>(Metadata.class, Metadata.EMPTY.getMainField());
    private final TaskInfo<Task<?>> taskInfo = new TaskInfo<>(null, new Label(), appProgressIndicator());
    private final Anim hideInfo = new Anim(at -> setScaleXY(taskInfo.getProgress(),at*at))
                                      .dur(millis(500)).intpl(reverse(new ElasticInterpolator()));
    private Output<Metadata> out_sel;

    @IsConfig(name = "Table orientation", info = "Orientation of the table.")
    public final Vo<NodeOrientation> orient = new Vo<>(APP.ui.getTableOrient());
    @IsConfig(name = "Zeropad numbers", info = "Adds 0s for number length consistency.")
    public final Vo<Boolean> zeropad = new Vo<>(APP.ui.getTableZeropad());
    @IsConfig(name = "Search show original index", info = "Show unfiltered table item index when filter applied.")
    public final Vo<Boolean> orig_index = new Vo<>(APP.ui.getTableOrigIndex());
    @IsConfig(name = "Show table header", info = "Show table header with columns.")
    public final Vo<Boolean> show_header = new Vo<>(APP.ui.getTableShowHeader());
    @IsConfig(name = "Show table footer", info = "Show table controls at the bottom of the table. Displays menu bar and table items information.")
    public final Vo<Boolean> show_footer = new Vo<>(APP.ui.getTableShowFooter());
    @IsConfig(editable = EditMode.APP) @Constraint.FileType(FileActor.ANY)
    private File lastFile = null;
    @IsConfig(editable = EditMode.APP) @Constraint.FileType(FileActor.DIRECTORY)
    private File lastDir = null;

    @SuppressWarnings({"unchecked", "ConstantConditions"})
    public Library(Widget widget) {
        super(widget);
        root.setPrefSize(600.0, 850.0);

        out_sel = outputs.create(widget.id, "Selected", Metadata.class, null);
        onClose.plusAssign(Player.librarySelected.i.bind(out_sel));

        root.getChildren().add(table.getRoot());

        // table properties
        table.getSelectionModel().setSelectionMode(MULTIPLE);
        table.search.setColumn(TITLE);
        onClose.plusAssign(maintain(orient,table.nodeOrientationProperty()));
        onClose.plusAssign(maintain(zeropad,table.zeropadIndex));
        onClose.plusAssign(maintain(orig_index,table.showOriginalIndex));
        onClose.plusAssign(maintain(show_header,table.headerVisible));
        onClose.plusAssign(maintain(show_footer,table.footerVisible));

        // add progress indicator to bottom controls
        ((Pane)table.footerPane.getRight()).getChildren().addAll(taskInfo.getMessage(), taskInfo.getProgress());
        taskInfo.setVisible(false);

        // extend table items information
        table.items_info.setTextFactory((all, list) -> {
            double lengthMs = list.stream().mapToDouble(Metadata::getLengthInMs).sum();
            return DEFAULT_TEXT_FACTORY.invoke(all, list) + " - " + new Dur(lengthMs);
        });

        // add more menu items
        table.menuAdd.getItems().addAll(
            menuItem("Add files", e -> addFiles()),
            menuItem("Add directory", e -> addDirectory())
        );
        table.menuRemove.getItems().addAll(
            menuItem("Remove selected songs from library", e -> APP.db.removeItems(table.getSelectedItems())),
            menuItem("Remove all shown songs from library", e -> APP.db.removeItems(table.getItems())),
            menuItem("Remove all songs from library", e -> APP.db.removeItems(table.getItems())),
            menuItem("Remove missing songs from library", e -> removeInvalid())
        );

        // set up table columns
        table.setColumnFactory(f -> {
            TableColumn<Metadata,Object> c = new TableColumn<>(f.toString());
            c.setCellValueFactory(cf -> cf.getValue()==null ? null : new PojoV<>(f.getOf(cf.getValue())));
            c.setCellFactory(f==(Metadata.Field) RATING
                ? (Callback) APP.getRatingCell().getValue()
                : column -> table.buildDefaultCell(f)
            );
            return c;
        });
        // maintain rating column cell style
        onClose.plusAssign(APP.getRatingCell().maintain(cf -> table.getColumn(RATING).ifPresent(c -> c.setCellFactory((Callback)cf))));   // maintain playing item css
        // let resizing as it is
        table.setColumnResizePolicy(resize -> {
            boolean b = UNCONSTRAINED_RESIZE_POLICY.call(resize);
            // resize index column
            table.getColumn(ColumnField.INDEX).ifPresent(i -> i.setPrefWidth(table.computeIndexColumnWidth()));
            return b;
        });

        table.getDefaultColumnInfo();   // TODO remove (this triggers menu initialization)

        // restore column state
        String columnStateS = widget.properties.getS("columns");
        table.setColumnState(columnStateS!=null ? TableColumnInfo.fromString(columnStateS) : table.getDefaultColumnInfo());

        // row behavior
        table.setRowFactory(tbl -> new ImprovedTableRow<Metadata>() {{
                onLeftDoubleClick((r,e) -> PlaylistManager.use(pl->pl.setNplayFrom(table.getItems(), r.getIndex())));
                onRightSingleClick((r,e) -> {
                    // prep selection for context menu
                    if (!r.isSelected())
                        tbl.getSelectionModel().clearAndSelect(r.getIndex());

                    contextMenu.show(MetadataGroup.groupOfUnrelated(table.getSelectedItemsCopy()), table, e);
                });
                styleRuleAdd(PC_PLAYING, m -> Player.playingItem.get().same(m));
            }}
        );
        onClose.plusAssign(Player.playingItem.onUpdate(o -> table.updateStyleRules()));   // maintain playing item css

        // maintain outputs
        table.getSelectionModel().selectedItemProperty().addListener((o,ov,nv) -> out_sel.setValue(nv));

        // key actions
        table.setOnKeyPressed(e -> {
            // play selected
            if (e.getCode() == ENTER) {
                if (!table.getSelectionModel().isEmpty()) {
                    PlaylistManager.use(pl -> pl.setNplayFrom(table.getItems(), table.getSelectionModel().getSelectedIndex()));
                }
            }
            // delete selected
            if (e.getCode() == DELETE) {
                APP.db.removeItems(table.getSelectedItems());
            }
        });

        // drag&drop from
        table.setOnDragDetected(e -> {
            if (e.getButton() == PRIMARY && !table.getSelectedItems().isEmpty()
                    && table.isRowFull(table.getRowS(e.getSceneX(), e.getSceneY()))) {
                Dragboard db = table.startDragAndDrop(COPY);
                DragUtil.setItemList(table.getSelectedItemsCopy(),db,true);
            }
            e.consume();
        });

        // prevent volume change
        table.setOnScroll(Event::consume);

        // update library comparator
        maintain(table.itemsComparator, APP.db.getLibraryComparator());


    }

    @Override
    public Collection<Config<Object>> getFields() {
        widget.properties.put("columns", table.getColumnState().toString());    // store column state
        return super.getFields();
    }

    /** Converts items to Metadata using {@link Item#toMeta()} (using no I/O) and displays them in the table. */
    @Override
    public void read(List<? extends Item> items) {
        if (items==null) return;
        table.setItemsRaw(map(items,Item::toMeta));
    }

    @IsInput("To display")
    public void setItems(List<? extends Metadata> items) {
        if (items==null) return;
        table.setItemsRaw(items);
    }

    private void addDirectory() {
        Window w = root.getScene().getWindow();
        FileChooser.ExtensionFilter ef = AudioFileFormat.filter(Use.APP);
        chooseFile("Add folder to library", FileType.DIRECTORY, lastDir, w, ef)
            .ifOk(file -> {
                APP.actionPane.show(file);
                lastDir = file.getParentFile();
            });
    }

    private void addFiles() {
        Window w = root.getScene().getWindow();
        FileChooser.ExtensionFilter ef = AudioFileFormat.filter(Use.APP);
        chooseFiles("Add files to library", lastFile, w, ef)
            .ifOk(files -> {
                APP.actionPane.show(files);
                lastFile = getCommonRoot(files);
            });
    }

    private void removeInvalid() {
        Task<Void> t = MetadataReader.buildRemoveMissingFromLibTask();
        fut(t)
            .useBy(FX, taskInfo::showNbind)
            .useBy(Task::run)
            .thenWait(seconds(5))
            .useBy(FX, it -> hideInfo.playOpenDo(runnable(taskInfo::hideNunbind)));
    }

}

File: app/widgets/library/Library.kt
package library

import javafx.scene.control.SelectionMode.MULTIPLE
import javafx.scene.control.TableView.UNCONSTRAINED_RESIZE_POLICY
import javafx.scene.input.KeyCode.DELETE
import javafx.scene.input.KeyCode.ENTER
import javafx.scene.input.KeyEvent.KEY_PRESSED
import javafx.scene.input.MouseButton.PRIMARY
import javafx.scene.input.MouseEvent.DRAG_DETECTED
import javafx.scene.input.TransferMode.ANY
import javafx.util.Callback
import sp.it.pl.audio.Song
import sp.it.pl.audio.playlist.PlaylistManager
import sp.it.pl.audio.tagging.Metadata
import sp.it.pl.audio.tagging.Metadata.Field.RATING
import sp.it.pl.audio.tagging.Metadata.Field.TITLE
import sp.it.pl.audio.tagging.MetadataGroup
import sp.it.pl.audio.tagging.removeMissingFromLibTask
import sp.it.pl.ui.objects.contextmenu.ValueContextMenu
import sp.it.pl.ui.objects.rating.RatingCellFactory
import sp.it.pl.ui.objects.table.FilteredTable
import sp.it.pl.ui.objects.table.ImprovedTable.PojoV
import sp.it.pl.ui.objects.table.buildFieldedCell
import sp.it.pl.ui.objects.tablerow.SpitTableRow
import sp.it.pl.layout.Widget
import sp.it.pl.main.WidgetTags.LIBRARY
import sp.it.pl.layout.controller.SimpleController
import sp.it.pl.layout.feature.SongReader
import sp.it.pl.main.APP
import sp.it.pl.main.AppProgress
import sp.it.pl.main.audioExtensionFilter
import sp.it.pl.main.emScaled
import sp.it.pl.main.setSongsAndFiles
import sp.it.pl.main.showConfirmation
import sp.it.pl.ui.nodeinfo.ListLikeViewInfo.Companion.DEFAULT_TEXT_FACTORY
import sp.it.util.access.fieldvalue.ColumnField
import sp.it.util.async.runNew
import sp.it.util.collections.materialize
import sp.it.util.conf.Config
import sp.it.util.conf.EditMode
import sp.it.util.conf.cn
import sp.it.util.conf.def
import sp.it.util.conf.noUi
import sp.it.util.conf.only
import sp.it.util.file.FileType.DIRECTORY
import sp.it.util.file.Util.getCommonRoot
import sp.it.util.functional.asIs
import sp.it.util.functional.invoke
import sp.it.util.functional.net
import sp.it.util.functional.orNull
import sp.it.util.reactive.consumeScrolling
import sp.it.util.reactive.on
import sp.it.util.reactive.onEventDown
import sp.it.util.reactive.sync
import sp.it.util.reactive.sync1IfInScene
import sp.it.util.system.chooseFile
import sp.it.util.system.chooseFiles
import sp.it.util.text.pluralUnit
import sp.it.util.ui.dsl
import sp.it.util.ui.lay
import sp.it.util.ui.prefSize
import sp.it.util.ui.pseudoclass
import sp.it.util.ui.x
import sp.it.util.units.millis
import sp.it.util.units.toHMSMs
import java.io.File
import sp.it.pl.ui.objects.table.TableColumnInfo as ColumnState
import mu.KLogging
import sp.it.pl.layout.WidgetCompanion
import sp.it.pl.main.Css.Pseudoclasses.played
import sp.it.pl.main.HelpEntries
import sp.it.pl.main.IconFA
import sp.it.pl.main.IconUN
import sp.it.pl.main.Widgets.SONG_TABLE_NAME
import sp.it.pl.main.installDrag
import sp.it.pl.ui.pane.ShortcutPane.Entry
import sp.it.util.Sort
import sp.it.util.access.OrV.OrValue.Initial.Inherit
import sp.it.util.collections.setTo
import sp.it.util.conf.cOr
import sp.it.util.conf.defInherit
import sp.it.util.functional.asIf
import sp.it.util.text.keys
import sp.it.util.text.nameUi
import sp.it.util.ui.show
import sp.it.util.ui.tableColumn
import sp.it.util.units.version
import sp.it.util.units.year

class Library(widget: Widget): SimpleController(widget), SongReader {

   private val table = FilteredTable(Metadata::class.java, Metadata.EMPTY.getMainField())
   private val outputSelected = io.o.create<Metadata?>("Selected", null)
   private val inputItems = io.i.create<List<Metadata>>("To display", listOf()) { setItems(it) }

   val tableOrient by cOr(APP.ui::tableOrient, table.nodeOrientationProperty(), Inherit(), onClose)
      .defInherit(APP.ui::tableOrient)
   val tableZeropad by cOr(APP.ui::tableZeropad, table.zeropadIndex, Inherit(), onClose)
      .defInherit(APP.ui::tableZeropad)
   val tableOrigIndex by cOr(APP.ui::tableOrigIndex, table.showOriginalIndex,  Inherit(), onClose)
      .defInherit(APP.ui::tableOrigIndex)
   val tableShowHeader by cOr(APP.ui::tableShowHeader, table.headerVisible, Inherit(), onClose)
      .defInherit(APP.ui::tableShowHeader)
   val tableShowFooter by cOr(APP.ui::tableShowFooter, table.footerVisible, Inherit(), onClose)
      .defInherit(APP.ui::tableShowFooter)
   private var lastAddFilesLocation by cn<File>(APP.location.user).noUi()
      .def(name = "Last add songs browse location", editable = EditMode.APP)
   private var lastAddDirLocation by cn<File>(APP.location.user).only(DIRECTORY).noUi()
      .def(name = "Last add directory browse location", editable = EditMode.APP)

   init {
      root.prefSize = 850.emScaled x 600.emScaled
      root.lay += table.root

      // table properties
      table.selectionModel.selectionMode = MULTIPLE
      table.search.setColumn(TITLE)
      table.items_info.textFactory = { all, list -> DEFAULT_TEXT_FACTORY(all, list) + " - " + list.sumOf { it.getLengthInMs() }.millis.toHMSMs() }

      // add menu items
      table.menuAdd.dsl {
         item("Add files") { addFiles() }
         item("Add directory") { addDirectory() }
      }
      table.menuRemove.dsl {
         item("Remove selected songs from library (${keys(DELETE)})") { removeSongs(table.selectedItems) }
         item("Remove all shown songs from library") { removeSongs(table.items) }
         item("Remove all songs from library") { APP.db.removeAllSongs() }
         item("Remove missing songs from library") {
            val task = Song.removeMissingFromLibTask()
            runNew(task)
            AppProgress.start(task)
         }
      }

      // set up table columns
      table.setColumnFactory { field ->
         tableColumn<Metadata, Any?>(field.name()) {
            cellFactory = when (field) {
               RATING -> RatingCellFactory.asIs()
               else -> Callback { field.buildFieldedCell() }
            }
            cellValueFactory = Callback { it.value?.net { PojoV(field.getOf(it)) } }
         }
      }

      // column resizing
      table.columnResizePolicy = Callback {
         UNCONSTRAINED_RESIZE_POLICY(it).apply {
            table.getColumn(ColumnField.INDEX).orNull()?.prefWidth = table.computeIndexColumnWidth()
         }
      }

      // row behavior
      table.rowFactory = Callback { t ->
         SpitTableRow<Metadata>().apply {
            onLeftDoubleClick { r, _ -> PlaylistManager.use { it.setAndPlayFrom(table.items, r.index) } }
            onRightSingleClick { r, e ->
               if (!r.isSelected) t.selectionModel.clearAndSelect(r.index)
               ValueContextMenu<MetadataGroup>().apply {
                  setItemsFor(MetadataGroup.groupOfUnrelated(table.selectedItemsCopy))
                  show(table, e)
               }
            }
            styleRuleAdd(pseudoclass(played)) { APP.audio.playingSong.value.same(it) }
         }
      }
      APP.audio.playingSong.onUpdate { _, _ -> table.updateStyleRules() } on onClose

      table.defaultColumnInfo   // trigger menu initialization
      table.columnState = widget.properties["columns"].asIf<String>()?.let { ColumnState.fromString(it).orNull() } ?: table.defaultColumnInfo

      table.onEventDown(KEY_PRESSED, ENTER, false) {
         if (!it.isConsumed) {
            if (!table.selectionModel.isEmpty) {
               PlaylistManager.use { it.setAndPlayFrom(table.items, table.selectionModel.selectedIndex) }
               it.consume()
            }
         }
      }
      table.onEventDown(KEY_PRESSED, DELETE, false) {
         if (!it.isConsumed) {
            if (!table.selectionModel.isEmpty) {
               removeSongs(table.selectedItems)
               it.consume()
            }
         }
      }

      // drag % drop
      table.onEventDown(DRAG_DETECTED, PRIMARY, false) {
         if (!table.selectedItems.isEmpty() && table.isRowFull(table.getRowS(it.sceneX, it.sceneY))) {
            table.startDragAndDrop(*ANY).setSongsAndFiles(table.selectedItemsCopy)
            it.consume()
         }
      }
      // drag % drop - prevent parent
      root.installDrag(IconFA.PLUS, "Add to library", { true }, { true }, {})

      table.consumeScrolling()

      // sync outputs
      table.selectionModel.selectedItemProperty() sync { outputSelected.value = it } on onClose
      root.sync1IfInScene { inputItems.bindDefaultIf1stLoad(APP.db.songs) } on onClose

      // sync library comparator
      table.itemsComparator sync {
         val sorts = table.sortOrder.mapNotNull { c -> c.userData?.asIs<Metadata.Field<*>>()?.net { Sort.of(c) to it } }
         APP.audio.songOrderBys setTo sorts.map { it.second }
         APP.audio.songOrderSorts setTo sorts.map { it.first }
      } on onClose

   }

   override fun getConfigs(): Collection<Config<Any?>> {
      widget.properties["columns"] = table.columnState.toString()
      return super.getConfigs()
   }

   override fun read(songs: List<Song>) {
      inputItems.value = songs.map { it.toMeta() }
   }

   fun setItems(items: List<Metadata>?) {
      if (items==null) return
      table.setItemsRaw(items)
   }

   private fun addDirectory() = chooseFile("Add folder to library", DIRECTORY, lastAddDirLocation, root.scene.window).ifOk {
      APP.ui.actionPane.orBuild.show(it)
      lastAddDirLocation = it.parentFile
   }

   private fun addFiles() = chooseFiles("Add files to library", lastAddFilesLocation, root.scene.window, audioExtensionFilter()).ifOk {
      APP.ui.actionPane.orBuild.show(it)
      lastAddFilesLocation = getCommonRoot(it)
   }

   fun removeSongs(songs: List<Song>) {
      if (songs.isEmpty()) {
         showConfirmation("No songs to remove - selection is empty") {}
      } else {
         val songsM = songs.materialize()
         showConfirmation("Are you sure you want to remove ${"song".pluralUnit(songs.size)} from library?") {
            APP.db.removeSongs(songsM)
         }
      }
   }

   companion object: WidgetCompanion, KLogging() {
      override val name = SONG_TABLE_NAME
      override val description = "Table of songs"
      override val descriptionLong = "$description. Allows access to song database."
      override val icon = IconUN(0x2e2a)
      override val version = version(1, 0, 0)
      override val isSupported = true
      override val year = year(2015)
      override val author = "spit"
      override val contributor = ""
      override val tags = setOf(LIBRARY)
      override val summaryActions = HelpEntries.Table + listOf(
         Entry("Table row", "Plays item", "2x${PRIMARY.nameUi}"),
      )
   }

}